sdk/eventhubs/azure-messaging-eventhubs-checkpointstore-blob/src/main/java/com/azure/messaging/eventhubs/checkpointstore/blob/MetricsHelper.java

```diff
@@ -0,0 +1,188 @@
+    private final ConcurrentHashMap<String, TelemetryAttributes> checkpointSuccess = new ConcurrentHashMap<>();
+    private final ConcurrentHashMap<String, CurrentValue> seqNoSubscriptions = new ConcurrentHashMap<>();
+
+    private volatile boolean maxCapacityReached = false;
+
+    private final Meter meter;
+    private final LongGauge lastSequenceNumber;
+    private final LongCounter checkpointCounter;
+    private final boolean isEnabled;
+
+    MetricsHelper(MetricsOptions metricsOptions, MeterProvider meterProvider) {
+        if (areMetricsEnabled(metricsOptions)) {
+            this.meter = meterProvider.createMeter(LIBRARY_NAME, LIBRARY_VERSION, metricsOptions);
+            this.isEnabled = this.meter.isEnabled();
+        } else {
+            this.isEnabled = false;
+            this.meter = null;
+        }
+
+        if (isEnabled) {
+            this.lastSequenceNumber = this.meter.createLongGauge("messaging.eventhubs.checkpoint.sequence_number", "Last successfully checkpointed sequence number.", "seqNo");
+            this.checkpointCounter = this.meter.createLongCounter("messaging.eventhubs.checkpoints", "Number of checkpoints.", null);
+        } else {
+            this.lastSequenceNumber = null;
+            this.checkpointCounter = null;
+        }
+    }
+
+    void reportCheckpoint(Checkpoint checkpoint, String attributesId, boolean success) {
+        if (!isEnabled || !(lastSequenceNumber.isEnabled() && checkpointCounter.isEnabled())) {
+            return;
+        }
+
+        if (!maxCapacityReached && (seqNoSubscriptions.size() >= MAX_ATTRIBUTES_SETS || common.size() >= MAX_ATTRIBUTES_SETS)) {
+            LOGGER.error("Too many attribute combinations are reported for checkpoint metrics, ignoring any new dimensions.");
+            maxCapacityReached = true;
+        }
+
+        if (lastSequenceNumber.isEnabled() && success) {
+            updateCurrentValue(attributesId, checkpoint);
+        }
+
+        if (checkpointCounter.isEnabled()) {
+            TelemetryAttributes attributes = null;
+            if (success) {
+                attributes = getOrCreate(checkpointSuccess, attributesId, checkpoint, "ok");
+            } else {
+                attributes = getOrCreate(checkpointFailure, attributesId, checkpoint, "error");
+            }
+            if (attributes != null) {
+                checkpointCounter.add(1, attributes, Context.NONE);
+            }
+        }
+    }
+
+    private TelemetryAttributes getOrCreate(ConcurrentHashMap<String, TelemetryAttributes> source, String attributesId, Checkpoint checkpoint, String status) {
+        if (maxCapacityReached) {
+            return source.get(attributesId);
+        }
+
+        return source.computeIfAbsent(attributesId, i -> meter.createAttributes(createAttributes(checkpoint, status)));
+    }
+
+    private Map<String, Object> createAttributes(Checkpoint checkpoint, String status) {
+        Map<String, Object> attributesMap = new HashMap<>();
```

Comment:
Since we know we put at most 5 key-values, should we create the hash-map with that capacity?